### PR TITLE
AL-4757 when refresh InMemoryFileIndex, can recursiveDirChildrenFiles

### PR DESCRIPTION
**Pull request by jlfsdtc**

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 1]
```diff
@@ -91,10 +91,28 @@ class InMemoryFileIndex(
   }
 
   private def refresh0(): Unit = {
-    val files = listLeafFiles(rootPaths)
-    cachedLeafFiles =
-      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
-    cachedLeafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+    if (sparkSession.sessionState.conf.scanSubDirEnabled) {
+      val (files, discovered) = listLeafFilesScanSubDirEnable(rootPaths)
```
Comment (yugan91sh) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala:95]: private def refresh0(): Unit = {
    val files = listLeafFiles(rootPaths)
    cachedLeafFiles =
      new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
    cachedLeafDirToChildrenFiles = if (sparkSession.sessionState.conf.recursiveFileLookup) {
      recursiveDirChildrenFiles(files)
    } else {
      files.toArray.groupBy(_.getPath.getParent)
    }
    cachedPartitionSpec = null
  }

  private def recursiveDirChildrenFiles(files: mutable.LinkedHashSet[FileStatus])
  : Map[Path, Array[FileStatus]] = {
    val rootPathSet = rootPaths.toSet
    val reorganized = mutable.LinkedHashMap[Path, mutable.LinkedHashSet[FileStatus]]()
    files.foreach { f => // f is a file, not a directory.
      var parent = f.getPath.getParent
      while(parent != null && !rootPathSet.contains(parent)) {
        reorganized.getOrElseUpdate(parent, mutable.LinkedHashSet[FileStatus]()).add(f)
        parent = parent.getParent
      }
    }
    reorganized.mapValues(_.toArray).toMap
  }

**Review by yugan91sh — CHANGES_REQUESTED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 1 above)
Comment (yugan91sh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala:95]: root path:
/user/hive/ssb.db/lineorders/


/user/hive/ssb.db/lineorders/part1=202210/a1.parquet
/user/hive/ssb.db/lineorders/part1=202210/a2.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/b1.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/b2.parquet
/user/hive/ssb.db/lineorders/part1=202210/z1.parquet

before change:

/user/hive/ssb.db/lineorders/part1=202210/ -> a1.parquet,a2.parquet,z1.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/ -> b1.parquet,b2.parquet

after change:

/user/hive/ssb.db/lineorders/part1=202210/ -> a1.parquet,a2.parquet,b1.parquet,b2.parquet,z1.parquet
/user/hive/ssb.db/lineorders/part1=202210/tez1/ -> b1.parquet,b2.parquet

-----


/user/hive/ssb.db/lineorders/part1=202211/tez1/b1.parquet
/user/hive/ssb.db/lineorders/part1=202211/tez1/b2.parquet

before change:
/user/hive/ssb.db/lineorders/part1=202211/tez1/ -> b1.parquet,b2.parquet

after change:
/user/hive/ssb.db/lineorders/part1=202211/ -> b1.parquet,b2.parquet
/user/hive/ssb.db/lineorders/part1=202211/tez1/ -> b1.parquet,b2.parquet

----

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1361,6 +1361,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val RECURSIVE_FILE_LOOKUP =
+    buildConf("spark.sql.sources.recursiveFileLookup")
+      .doc("If true, Spark will not fetch the block locations for each file on " +
```
Comment (yugan91sh) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1366]: 这个说明和配置对不上吧，是 spark.sql.sources.ignoreDataLocality 的说明

**Review by yugan91sh — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (jlfsdtc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1366]: 这个还没改, 还在测试功能

**Comment by jlfsdtc** [comment]
image

**Review by zheniantoushipashi — APPROVED** [review]

**Review by yugan91sh — APPROVED** [review]

**Pull request merged**